### ISSUE TITLE
[SolYul] Support all Artefacts

### ISSUE THREAD

**Issue by chriseth** [ethereum/solidity#11494]
When compiling Solidity via yul, it should be possible to output the IR as a text file and re-start the compilation from there to obtain exactly the same bytecode. For that, we need to serialize the compiler state for each contract into comments in the Yul code and re-parse that again.

Sub-tasks:
 - [1] store Solidity source locations in comments when generating the IR https://github.com/ethereum/solidity/pull/11450
 - [2] store additional metadata in IR comments at contract level (metadata, storage layout, natspec, generated sources?)
 - [3] when re-parsing the Yul IR, we for now do not change the way source locations are stored in the Yul AST nodes, but we do process all doxygen comments and store tag-value pairs (strings) and associate them with the Yul AST node addresses that come after them.
 - [4] the data generated in [2] is extracted from these strings again. Furthermore, we do a pass over the Yul AST to replace the source locations extracted from the strings.
 - [5] after the optimizer has run, we should be able to produce the same set of data as in [1] and [2], but from the structured data instead of just during code generation.

In terms of code, the task here is to extend `CompilerStack::generateEVMFromIR` so that it can generate the same information as `CompilerStack::compileContract` - it should be able to fill most of the fields of `CompilerStack::Contract`.
At the same time, `StandardCompiler::compileYul` should do the same thing as `CompilerStack::generateEVMFromIR`, when the input Yul source is properly annotated.

**Comment by chriseth** [comment]
When the compiler is run on the IR generated in [2], it has to detect from comments whether this is hand-written yul or generated from Solidity. It would be nice to have a mostly uniform process that only generates more information when the relevant comments are present - this is mostly done in step [4]. There should be a contract-level comment that tells how to interpret the comments that follow.

Note that it is also possible to start with hand-written Yul, run it through the optimizer and output it again. This output should have source location comments which do not refer to a solidity file, but are still "origin" source locations.

Step [4] has to be able to distinguish these cases and apply the two source locations properly for all combinations.

**Comment by Marenz** [comment]
related: https://github.com/ethereum/solidity/issues/11086

**Comment by cameel** [comment]
The locations are blocked by #11086 (we need support for `@use-src` annotation) so I'd like to add support for other artifacts. @chriseth, is the format for storing stack layout in annotations already decided? Do we actually want to dump it into the Yul file?

**Comment by cameel** [comment]
Looks like `@use-src` is done after all (#11594). From @chriseth:

> the only thing left is the function entry points. I have a pending pr that proposes a way to do it, but not sure if that is a good way. Currently, we don't export stack layout in any way.
and the source part is actually finished, I hope. One thing that is not decided yet but also not required is if we want to have source annotations for the intermediate representation and not just the solidity source code
oh and another thing that would be nice is to print the source snippet (in AsmPrinter) at the location like we do in the --asm output, instead of just the numbers.

The PR related to entry points is #11661.

I have created an issue for the source snippets: #11858.

**Comment by cameel** [comment]
Looks like this task is not complete yet.  The output of two-stage compilation is still different than from the single-stage one. Reported as a separate issue not to spam this one too much with details: #11875.

In short: the code snippets next to code locations do not show up (in EVM assembly not Yul) ~and the metadata seems different despite there being no differences in file names and not extra flags beyond `--experimental-via-ir`~.

EDIT: Metadata differences were caused by that flag. I think it might still be a problem for reproducibility if people are not aware of this. Maybe we should remove that flag from metadata?

**Comment by cameel** [comment]
Looks like #11875 is not a problem after all. The snippets are rightly missing because the second step no longer has access to the original Solidity source. We could propagate them from Yul by storing them in debug data but that could cause inconsistencies and we'd first need to decide how to deal with them.

As for the flag in metadata, it's not a problem, but `--ir` should be automatically adding it: #11876.

We should also have a test checking that results of one-step and two-step compilation match: #11877.

**Comment by chriseth** [comment]
I think this is done, right? We only need to check that the two-step process does the same as the single-step process.

**Comment by cameel** [comment]
There's also #11858 (the part about a setting for disabling debug info) but that already has a PR waiting for review.
